File: src/BaseHmiTypes/Converters/AutomationXml/HmiAutomationXmlConverter.cs
using System.Globalization;
using System.Text;
using System.Xml;
using BaseHmiTypes.Common;
using BaseHmiTypes.Cycles;
using BaseHmiTypes.Screens.Base;
using BaseHmiTypes.Scripts;
using BaseHmiTypes.TextGraphicLists;

namespace BaseHmiTypes.Converters.AutomationXml;

public static class HmiAutomationXmlConverter
{
    public static string ToAutomationXml(params IHmiObject[] objects)
    {
        using var ms = new MemoryStream();
        using (var writer = XmlWriter.Create(ms, new XmlWriterSettings { Indent = true, Encoding = Encoding.UTF8 }))
        {
            var id = 0;
            writer.WriteStartElement("Document");
            writer.WriteStartElement("Engineering");
            writer.WriteAttributeString("version", "V21");
            writer.WriteEndElement();

            foreach (var obj in objects)
            {
                if (obj is HmiCycle cycle)
                    HmiCycleToAutomationXmlConverter.WriteAutomationXml(writer, cycle, ref id);
                else if (obj is HmiScript script)
                    HmiScriptToAutomationXmlConverter.WriteAutomationXml(writer, script, ref id);
                else if (obj is HmiTextList textList)
                    HmiTextListToAutomationXmlConverter.WriteAutomationXml(writer, textList, ref id);
                else if (obj is HmiScreenBase screen)
                    HmiScreenToAutomationXmlConverter.WriteAutomationXml(writer, screen, ref id);
            }

            writer.WriteEndElement();
        }

        ms.Position = 0;
        using var reader = new StreamReader(ms);
        return reader.ReadToEnd();
    }

    public static void WriteAutomationXml(HmiMultilingualText multiLanguageText, XmlWriter writer, ref int id, string compositionName)
    {
        writer.WriteStartElement("MultilingualText");
        writer.WriteAttributeString("ID", "0x" + (id++).ToString("X", CultureInfo.InvariantCulture));
        writer.WriteAttributeString("CompositionName", compositionName);
        writer.WriteStartElement("ObjectList");
        foreach (var text in multiLanguageText.Texts)
        {
            writer.WriteStartElement("MultilingualTextItem");
            writer.WriteAttributeString("ID", "0x" + (id++).ToString("X", CultureInfo.InvariantCulture));
            writer.WriteAttributeString("CompositionName", "Items");
            writer.WriteStartElement("AttributeList");
            writer.WriteStartElement("Culture");
            if (text.Key >= 0)
                writer.WriteString(new CultureInfo(text.Key).Name);
            writer.WriteEndElement();
            writer.WriteStartElement("Text");
            writer.WriteString(text.Value);
            writer.WriteEndElement();
            writer.WriteEndElement();
            writer.WriteEndElement();
        }

        writer.WriteEndElement();
        writer.WriteEndElement();
    }

}

File: typescript/base-hmi-types/src/converters/automation-xml/HmiAutomationXmlConverter.ts
import { IHmiObject } from "../../IHmiObject.js";
import { HmiMultilingualText } from "../../common/HmiMultilingualText.js";
import { HmiCycle } from "../../cycles/HmiCycle.js";
import { HmiScreenBase } from "../../screens/base/HmiScreenBase.js";
import { HmiScript } from "../../scripts/HmiScript.js";
import { HmiTextList } from "../../text-graphic-lists/HmiTextList.js";
import { HmiCycleToAutomationXmlConverter } from "./HmiCycleToAutomationXmlConverter.js";
import { HmiScreenToAutomationXmlConverter } from "./HmiScreenToAutomationXmlConverter.js";
import { HmiScriptToAutomationXmlConverter } from "./HmiScriptToAutomationXmlConverter.js";
import { HmiTextListToAutomationXmlConverter } from "./HmiTextListToAutomationXmlConverter.js";
import { XmlWriter } from "./XmlWriter.js";

export class HmiAutomationXmlConverter {
  static toAutomationXml(...objects: IHmiObject[]): string {
    const writer = new XmlWriter();
    const id = { value: 0 };
    writer.writeStartElement("Document");
    writer.writeStartElement("Engineering");
    writer.writeAttributeString("version", "V21");
    writer.writeEndElement("Engineering");

    for (const obj of objects) {
      if (obj instanceof HmiCycle) {
        HmiCycleToAutomationXmlConverter.writeAutomationXml(writer, obj, id);
      } else if (obj instanceof HmiScript) {
        HmiScriptToAutomationXmlConverter.writeAutomationXml(writer, obj, id);
      } else if (obj instanceof HmiTextList) {
        HmiTextListToAutomationXmlConverter.writeAutomationXml(writer, obj, id);
      } else if (obj instanceof HmiScreenBase) {
        HmiScreenToAutomationXmlConverter.writeAutomationXml(writer, obj, id);
      }
    }

    writer.writeEndElement("Document");
    return writer.toString();
  }

  static writeMultilingualText(multiLanguageText: HmiMultilingualText, writer: XmlWriter, id: { value: number }, compositionName: string): void {
    writer.writeStartElement("MultilingualText");
    writer.writeAttributeString("ID", `0x${(id.value++).toString(16).toUpperCase()}`);
    writer.writeAttributeString("CompositionName", compositionName);
    writer.writeStartElement("ObjectList");
    for (const [culture, text] of multiLanguageText.texts) {
      writer.writeStartElement("MultilingualTextItem");
      writer.writeAttributeString("ID", `0x${(id.value++).toString(16).toUpperCase()}`);
      writer.writeAttributeString("CompositionName", "Items");
      writer.writeStartElement("AttributeList");
      writer.writeElementString("Culture", culture >= 0 ? String(culture) : "");
      writer.writeElementString("Text", text);
      writer.writeEndElement("AttributeList");
      writer.writeEndElement("MultilingualTextItem");
    }
    writer.writeEndElement("ObjectList");
    writer.writeEndElement("MultilingualText");
  }
}
